File: Programmers/python/level1/체육복.py


def solution(n, lost, reserve):
    answer = n
    lost.sort()
    reserve.sort()
    # 자신에게 여분이 있는 학생
    for i in range(1, n + 1):
        if i in lost:
            if i in reserve:
                reserve.remove(i)
                lost.remove(i)

    # 빌려야 하는 학생
    for l in lost:
        if l - 1 in reserve:
            reserve.remove(l - 1)
        elif l + 1 in reserve:
            reserve.remove(l + 1)
        else:
            answer -= 1
    return answer

# 다른 사람 풀이
# def solution(n, lost, reserve):
#     answer = 0
#     for i in range(1, n+1):
#         if i not in lost: #안 잃어버린 학생
#             answer += 1
#         else:
#             if i in reserve: #잃어버렸지만 여분도 있는 학생
#                 answer += 1
#                 reserve.remove(i)
#                 lost.remove(i)

#     for i in lost: #잃어버리고 여분도 없어서 빌려야 하는 학생
#         if i-1 in reserve:
#             answer += 1
#             reserve.remove(i-1)

#         elif i+1 in reserve:
#             answer += 1
#             reserve.remove(i+1)

#     return answer
n = 5
lost = [2,4]
reserve = [1,3,5]
print(solution(n, lost, reserve))

File: Programmers/Java_lan/level1/체육복.java
import java.util.*;

public class 체육복 {
    public static void main(String[] args) {
        int n = 5;
        int[] lost = {2, 4};
        int[] reserve = {1, 3, 5};
        System.out.println(solution(n, lost, reserve));
    }
    
    public static int solution(int n, int[] lost, int[] reserve) {
        int answer = n - lost.length;
        Arrays.sort(lost);
        Arrays.sort(reserve);
        
        // 여벌을 가져온 학생이 체육복을 도난당한 경우    
        for(int i = 0; i < lost.length; i++) {
            for(int j = 0; j < reserve.length; j++) {
                if(lost[i] == reserve[j]) {
                    answer++;
                    lost[i] = -1;
                    reserve[j] = -1;
                    break;
                }
            }
        }
        
        // 도난당한 학생에게 체육복을 빌려주는 경우
        for(int i = 0; i < lost.length; i++) {
            for(int j = 0; j < reserve.length; j++) {
                if((lost[i] - 1 == reserve[j]) || (lost[i] + 1 == reserve[j])) {
                    answer++;
                    reserve[j] = -1;
                    break;
                }
            }
        }
        return answer;
    }
}
